### PR TITLE
[FIX] some memory leaks with physics objects (ammo.js)

### PR DESCRIPTION
**Pull request by Maksims**
I screw up https://github.com/playcanvas/engine/pull/1742 with git, still messing up sometimes :)
So closing other PR, and working off of this.

Related forum post: https://forum.playcanvas.com/t/memory-leaks-of-underlying-physics-engine/10528

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

**Comment by Maksims** [comment]
**This can be tested and merged.**

I've done extensive testing on some projects I could find. If you have physics game with different mechanics, please provide links for me to test.

Some results:

This is example provided on [forum](https://forum.playcanvas.com/t/memory-leaks-of-underlying-physics-engine/10528). Here is image, showing physics memory difference with 10 scene changes (with current engine):
image

And here is with updated:
image

Also notable [project](https://playcanvas.com/project/352761/overview/chickens-day-off) where this PR makes big difference (played till score 20):
Before:
image

After:
image

File: src/framework/components/rigid-body/system.js [hunk 1]
```diff
@@ -239,10 +239,14 @@ Object.assign(pc, function () {
         onRemove: function (entity, data) {
             if (data.body) {
                 this.removeBody(data.body);
+
                 Ammo.destroy(data.body);
-            }
 
-            data.body = null;
+                var motionState = data.body.getMotionState();
+                delete Ammo.btMotionState.__cache__[motionState.ptr];
```
Comment (willeastcott) [inline src/framework/components/rigid-body/system.js]: This looks odd. Can you add code comments explaining the requirement for the above 2 lines please?

File: src/framework/components/collision/trigger.js [hunk 2]
```diff
@@ -46,14 +46,21 @@ Object.assign(pc, function () {
...
                 startTransform.setIdentity();
-                startTransform.getOrigin().setValue(pos.x, pos.y, pos.z);
+                var origin = startTransform.getOrigin();
+                origin.setValue(pos.x, pos.y, pos.z);
                 startTransform.setRotation(ammoQuat);
 
                 var motionState = new Ammo.btDefaultMotionState(startTransform);
                 var bodyInfo = new Ammo.btRigidBodyConstructionInfo(mass, motionState, shape, localInertia);
 
+                delete Ammo.btDefaultMotionState.__cache__[motionState.ptr];
```
Comment (willeastcott) [inline src/framework/components/collision/trigger.js]: This looks odd. Can you add code comments explaining the requirement for the above line please?

File: src/framework/components/collision/system.js [hunk 3]
```diff
@@ -400,6 +430,19 @@ Object.assign(pc, function () {
...
 
             pc.CollisionSystemImpl.prototype.updateTransform.call(this, component, position, rotation, scale);
+        },
+
+        remove: function(entity, data) {
+            if (data.shape) {
+                var numShapes = data.shape.getNumChildShapes();
+                for(var i = 0; i < numShapes; i++) {
+                    var shape = data.shape.getChildShape(i);
+                    delete Ammo.btBvhTriangleMeshShape.__cache__[shape.ptr];
```
Comment (willeastcott) [inline src/framework/components/collision/system.js]: Can you add code comments explaining the requirement for the above line please?

File: src/framework/components/rigid-body/system.js (hunk 1 above)
Comment (Maksims) [reply src/framework/components/rigid-body/system.js]: When btRigidBody is created, it creates btMotionState for it. And when deleting btRigidBody, it does not deletes its motion state. It is weird, does it reuses them internally - I'm not sure. But deleting it using `Ammo.destroy` will lead to errors. After further testing, just removing reference from cache - does the trick.
I did not went too far, profiling WebAssembly memory, I would assume they still have some objects references stored inside, unfortunately.

After a lot of testing, this has not brought any issue, and cleared JS references to it.

File: src/framework/components/collision/system.js (hunk 3 above)
Comment (Maksims) [reply src/framework/components/collision/system.js]: I might explore slightly other option, is to cast btShape object into btBvhTriangleMeshShape, and then call Ammo.destroy on it.
Destroying compound shape, does not destroys its children.

File: src/framework/components/rigid-body/system.js (hunk 1 above)
Comment (willeastcott) [reply src/framework/components/rigid-body/system.js]: Is it worth opening an issue on the Ammo repo and seeing if Alon Zakai has any thoughts?

File: src/framework/components/collision/system.js (hunk 3 above)
Comment (willeastcott) [reply src/framework/components/collision/system.js]: 👍

File: src/framework/components/rigid-body/system.js (hunk 1 above)
Comment (Maksims) [reply src/framework/components/rigid-body/system.js]: Removed the delete bit, and added an issue on ammo.js

File: src/framework/components/collision/trigger.js (hunk 2 above)
Comment (Maksims) [reply src/framework/components/collision/trigger.js]: Removed this bit.

File: src/framework/components/collision/system.js (hunk 3 above)
Comment (Maksims) [reply src/framework/components/collision/system.js]: Removed this bit.

**Comment by Maksims** [comment]
Removed `delete .. __cache__` bits, as they do seems hacky.
I would assume internally in ammo.js memory stack, related objects are removed, it is only references in js __cache__ objects are left, and they are usually tiny.

Also created an issue related to this: https://github.com/kripken/ammo.js/issues/284

**Pull request merged**